[IMP] Menu Service

hokolomopo opened the pull request:
## Description:

Add a menu service to handle the menus, so only one menu can be opened at a time.

Odoo task ID : [3142704](https://www.odoo.com/web#id=3142704&action=333&active_id=2328&model=project.task&view_type=form&cids=1&menu_id=4720)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_lt("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/o-spreadsheet/pull/1969).

hokolomopo commented on src/helpers/menu_service.ts:
```diff
@@ -0,0 +1,56 @@
+    return this.activeMenu?.id;
+  }
+
+  getCurrentMenuProps(): MenuProps | undefined {
+    const props: any = { ...this.activeMenu, depth: 1 };
+    if (!props) return undefined;
+    delete props["id"];
+    return props;
```
Typing sucks

LucasLefevre commented on src/components/spreadsheet/spreadsheet.ts:
```diff
@@ -174,6 +176,7 @@ export class Spreadsheet extends Component<SpreadsheetProps, SpreadsheetChildEnv
       toggleSidePanel: this.toggleSidePanel.bind(this),
       _t: Spreadsheet._t,
       clipboard: this.env.clipboard || instantiateClipboard(),
+      menuService: new MenuService(() => this.model.trigger("update")),
```
```suggestion
      menuService: new MenuService(() => this.render(true)),
```
Is equivalent and makes more sense. It has nothing to do with model.
Could be optimized by taking advantage of owl reactivity

hokolomopo commented on src/helpers/menu_service.ts:
```diff
@@ -0,0 +1,70 @@
+
+  unregisterMenu(id: string | undefined) {
+    if (id && this.activeMenu?.id === id) {
+      // call onClose after setting activeMenu to undefined to avoid possible infinite recursion
```
The problem occurs if you call `mernuService.unregisterMenu()` from the callback `onClose` of the menu. I feel like it's a very easy mistake to do, and preventing it from making everything crash doesn't cost much.

hokolomopo committed:
[IMP] menus:  add a menu service to manage menus

This commit create a menu service to handle the different menu. This
has 2 major advantages :
- the service is global, so opening a menu in a component will close
    the menu in the others components
- we don't have to manage a `MenuState` in every component that use a
    menu

Odoo task 3142704

hokolomopo commented on src/components/grid_popover/grid_popover.ts:
```diff
@@ -9,13 +10,32 @@ interface Props {
+        }
+      }
+      this.wasMenuOpen = isMenuOpen;
+    });
```
Maybe the persistent popover should be also registered in the menu service instead of having interaction between them and the menuService here ?

The main drawback is that the typing of the menuService would becomes more complex, and it would be harder to use it.

hokolomopo committed:
close menus on unmount

hokolomopo commented on tests/components/figure.test.ts:
```diff
@@ -423,14 +433,17 @@ describe("figures", () => {
       });
 
       test("Can open context menu on right click", async () => {
-        triggerMouseEvent(".o-figure", "contextmenu");
```
The were opening the grid context menu lol

LucasLefevre commented on src/helpers/menu_service.ts:
```diff
@@ -0,0 +1,71 @@
+
+  constructor() {}
+
+  registerMenu(menu: RegisterMenuArgs): UID {
```
```suggestion
  openMenu(menu: RegisterMenuArgs): UID {
```

LucasLefevre commented on src/components/grid_popover/grid_popover.ts:
```diff
@@ -9,13 +10,32 @@ interface Props {
+  private wasMenuOpen = this.menuService.hasOpenMenu();
+
+  setup() {
+    onPatched(() => {
```
`onWillUpdateProps` to avoid a useless rendering ? 
`onWillUpdateProps` is called *before* while `onPatched` is called *after*

hokolomopo replied:
Dispatching a command in `onWillUpdateProps` makes everything crash :x

LucasLefevre reviewed (commented):
I think a shared ("global") state is the good direction to solve this kind of problem.
It's the same problem as selection input and grid popover (currently UI plugins to share their state).

Several comments:

I wouldn't use blindly the `env` directly. For several reasons:
- it couples components to the env
- we risk conflicts with existing stuff in the env

The need to call `onWillUnmount` each time is error prone. We will definitely forget one day or another.

Now, given the two points above, a hook could solve both problems.
What do you think?

```ts

function useMenu() {
  const env = useEnv();
  const menu = env.__spreadsheet__.menuService;
  onWillUnmount(() => ...);
  return {
    open(...) {
      menu.open(...)
    },
    close() {
      ...
    },
  }
}
```

I didn't look into the details. Only the general idea, which I think I like :slightly_smiling_face:

hokolomopo commented on src/helpers/menu_service.ts:
```diff
@@ -0,0 +1,71 @@
+    };
+    this.activeMenu = newMenu;
+
+    return newMenu.id;
```
We need some kind of way to identidy the opened menu for the toggles (eg. in the bottom bar sheet).

```
if(this.contextMenuId === this.env.menuService.getCurrentMenuId()) closeMenu()
else openMenu()
```

hokolomopo committed:
lul fixes
